File: src/engine/tdeeCalculator.ts
import { UserProfile, CalculatedTargets } from '../types/profile';
import { ReferenceProfile } from '../types/assessment';

export function calculateUserTargets(profile: UserProfile): CalculatedTargets {
  const { weightKg, heightCm, age, sex, activityLevel, goal } = profile;

  // 1. Calculate Basal Metabolic Rate (BMR) using Mifflin-St Jeor Equation
  let bmr = 10 * weightKg + 6.25 * heightCm - 5 * age;
  if (sex === 'male') {
    bmr += 5;
  } else if (sex === 'female') {
    bmr -= 161;
  } else {
    bmr -= 78; // neutral midpoint
  }

  // 2. Physical Activity Multiplier
  const activityMultipliers: Record<string, number> = {
    sedentary: 1.2,
    light: 1.375,
    moderate: 1.55,
    very_active: 1.725,
    extra_active: 1.9,
  };

  const multiplier = activityMultipliers[activityLevel] || 1.55;
  const tdee = Math.round(bmr * multiplier);

  // 3. Goal Calorie Adjustment
  let targetCalories = tdee;
  let proteinRatio = 0.2; // 20%
  let fatRatio = 0.28; // 28%
  let carbRatio = 0.52; // 52%
  let targetSodium = 2300;
  let maxSugar = 50;

  switch (goal) {
    case 'fat_loss':
      targetCalories = Math.max(1200, Math.round(tdee - 450));
      proteinRatio = 0.28; // Higher protein for satiety & muscle preservation
      fatRatio = 0.27;
      carbRatio = 0.45;
      maxSugar = 35;
      break;

    case 'muscle_gain':
      targetCalories = Math.round(tdee + 350);
      proteinRatio = 0.25; // High protein (approx 1.8-2.0g/kg)
      fatRatio = 0.25;
      carbRatio = 0.5;
      break;

    case 'low_sodium':
      targetCalories = tdee;
      targetSodium = 1500; // Strict heart-healthy sodium target
      break;

    case 'low_sugar':
      targetCalories = tdee;
      maxSugar = 25; // 25g strict sugar limit
      break;

    case 'maintain':
    case 'healthy_eating':
    default:
      targetCalories = tdee;
      break;
  }

  // Calculate gram targets
  // Protein: 4 kcal/g, Carbs: 4 kcal/g, Fat: 9 kcal/g
  const targetProtein = Math.round((targetCalories * proteinRatio) / 4);
  const targetFat = Math.round((targetCalories * fatRatio) / 9);
  const targetCarbs = Math.round((targetCalories * carbRatio) / 4);
  const targetSaturatedFat = Math.round((targetCalories * 0.08) / 9); // <8% of calories
  const targetFiber = Math.max(25, Math.round((targetCalories / 1000) * 14)); // 14g per 1000 kcal standard
  const targetSalt = Number((targetSodium * 0.0025).toFixed(1));

  return {
    bmr: Math.round(bmr),
    tdee,
    targetCalories,
    targetProtein,
    targetCarbs,
    targetSugars: maxSugar,
    targetFat,
    targetSaturatedFat,
    targetFiber,
    targetSodium,
    targetSalt,
  };
}

export function buildPersonalizedReferenceProfile(profile: UserProfile): ReferenceProfile {
  const targets = calculateUserTargets(profile);

  return {
    id: 'personalized',
    name: `Personalized (${profile.name || 'Your Profile'})`,
    regionOrAuthority: `Goal: ${profile.goal.replace('_', ' ').toUpperCase()}`,
    description: `Custom targets calibrated to ${profile.age}y, ${profile.weightKg}kg, and your ${profile.goal.replace('_', ' ')} goal.`,
    calories: targets.targetCalories,
    protein: targets.targetProtein,
    carbohydrates: targets.targetCarbs,
    sugars: targets.targetSugars,
    addedSugars: Math.round(targets.targetSugars * 0.6),
    fat: targets.targetFat,
    saturatedFat: targets.targetSaturatedFat,
    fiber: targets.targetFiber,
    sodium: targets.targetSodium,
    salt: targets.targetSalt,
  };
}

File: android/app/src/main/java/com/nutriscan/app/domain/calculator/TdeeCalculator.kt
package com.nutriscan.app.domain.calculator

import com.nutriscan.app.data.model.ActivityLevel
import com.nutriscan.app.data.model.DietaryGoal
import com.nutriscan.app.data.model.Gender
import com.nutriscan.app.data.model.UserProfile
import kotlin.math.max
import kotlin.math.roundToInt

data class DailyNutritionTargets(
    val bmr: Int,
    val tdee: Int,
    val calories: Int,
    val proteinGrams: Int,
    val carbsGrams: Int,
    val fatGrams: Int,
    val sugarMaxGrams: Int,
    val fiberGrams: Int,
    val sodiumMg: Int
)

object TdeeCalculator {

    fun calculate(profile: UserProfile): DailyNutritionTargets {
        // 1. Mifflin-St Jeor BMR
        var bmr = 10.0 * profile.weightKg + 6.25 * profile.heightCm - 5.0 * profile.age
        when (profile.gender) {
            Gender.MALE -> bmr += 5.0
            Gender.FEMALE -> bmr -= 161.0
            Gender.NEUTRAL -> bmr -= 78.0
        }

        // 2. Activity Multiplier
        val multiplier = when (profile.activityLevel) {
            ActivityLevel.SEDENTARY -> 1.2
            ActivityLevel.LIGHT -> 1.375
            ActivityLevel.MODERATE -> 1.55
            ActivityLevel.VERY_ACTIVE -> 1.725
            ActivityLevel.EXTRA_ACTIVE -> 1.9
        }

        val tdee = (bmr * multiplier).roundToInt()

        // 3. Goal Calorie Adjustment
        var targetCalories = profile.customCalories ?: tdee
        var proteinRatio = 0.20
        var fatRatio = 0.28
        var carbRatio = 0.52
        var maxSugar = 50
        var targetSodium = 2000

        if (profile.customCalories == null) {
            when (profile.dietaryGoal) {
                DietaryGoal.FAT_LOSS -> {
                    targetCalories = max(1200, tdee - 450)
                    proteinRatio = 0.28
                    fatRatio = 0.27
                    carbRatio = 0.45
                    maxSugar = 35
                }
                DietaryGoal.MUSCLE_GAIN -> {
                    targetCalories = tdee + 350
                    proteinRatio = 0.25
                    fatRatio = 0.25
                    carbRatio = 0.50
                }
                DietaryGoal.LOW_SODIUM -> {
                    targetCalories = tdee
                    targetSodium = 1500
                }
                DietaryGoal.LOW_SUGAR -> {
                    targetCalories = tdee
                    maxSugar = 25
                }
                DietaryGoal.MAINTAIN -> {
                    targetCalories = tdee
                }
            }
        }

        // Gram targets
        val targetProtein = ((targetCalories * proteinRatio) / 4.0).roundToInt()
        val targetFat = ((targetCalories * fatRatio) / 9.0).roundToInt()
        val targetCarbs = ((targetCalories * carbRatio) / 4.0).roundToInt()
        val targetFiber = max(25, ((targetCalories / 1000.0) * 14.0).roundToInt())

        return DailyNutritionTargets(
            bmr = bmr.roundToInt(),
            tdee = tdee,
            calories = targetCalories,
            proteinGrams = targetProtein,
            carbsGrams = targetCarbs,
            fatGrams = targetFat,
            sugarMaxGrams = maxSugar,
            fiberGrams = targetFiber,
            sodiumMg = targetSodium
        )
    }
}
